File: server/types/dataset.go
package types

import (
	orderedMap "github.com/wk8/go-ordered-map/v2"
)

type DataContentType = *orderedMap.OrderedMap[string, string] // "text" | "image"

type DatasetPayload struct {
	Key      string          `json:"key"`
	KeyLabel string          `json:"keyLabel"`
	Columns  DataContentType `json:"columns"`
}

//type DatasetPayloadMap = *orderedMap.OrderedMap[string, any]

type DatasetRow map[string]string

type DatasetWithRows struct {
	DatasetPayload
	ID   string       `json:"id,omitempty"`
	Rows []DatasetRow `json:"rows,omitempty"`
}

File: frontend/types/dataset.ts
export type DatasetPayload = {
    /** The primary column name used as a unique identifier for input data matching. */
    key: string;
    /** The custom display text used to represent the {@link key} across the user interface. */
    keyLabel: string;
    /**
     * The map of dataset column names and their data types.
     * Used for content validation.
     * If the type is "text" then it will be shown as a regular text.
     * If the type is "image" then it will be shown as an image.
     *
     * @remarks
     * Updating or adding keys in this property *requires* you to
     * manually update {@link columnKeys} immediately after to prevent stale data.
     */
    columns: DataContentType;
};
export type CachedDataset = DatasetPayload & {
    /**
     * The dataset column names.
     *
     * @remarks
     * This is dependent value derived from {@link columns}.
     * Do not update this directly without a corresponding change to {@link columns}.
     */
    columnKeys: DatasetRowKey[];
};
export type DatasetRow = Record<string, string>;
export type DatasetWithRows = DatasetPayload & { rows?: (DatasetRow | null)[] };

export type DatasetRowKey = keyof DatasetRow;
export type DatasetRowValue = DatasetRow[DatasetRowKey];
export type DataContentType<T = DatasetRow> = {
    [K in keyof T]: "text" | "image";
};
